pkg/rotator/rotator.go

```diff
@@ -437,6 +437,9 @@ func injectCertToApiService(apiService *unstructured.Unstructured, certPem []byt
 	if !found {
 		return errors.New("`spec` field not found in APIService")
 	}
+	if err := unstructured.SetNestedField(apiService.Object, false, "spec", "insecureSkipTLSVerify"); err != nil {
+		return err
+	}
```

Comment:
I'm curious to get some of the maintainers thoughts on whether this actually has to happen here, in cert-controller. If `cert-controller` is going to be used with APIService, presumably we are going to inject a `caBundle` so why not set this field outside of `cert-controller` to begin with ?

Reply:
The problem we have faced with is that when you set `insecureSkipTLSVerify: true` , you have to set explicitly `insecureSkipTLSVerify: false` to remove the value. 
As `insecureSkipTLSVerify: false` is the default value, k8s api removes it automatically and CD tools like ArgoCD/Flux enter on permanent out-of-sync status.
https://github.com/kedacore/charts/issues/550

In this scenario, we have some options:
- Adding a step to remove the field manually (breaking the helm upgrade process)
- Adding explicitly the value in helm chart, producing out-of-sync on CD tools
- Updating `insecureSkipTLSVerify` before setting `caBundle`

Maybe I've missed something and there is any other option

Reply:
I'm a bit confused here. You're saying that having `insecureSkipTLSVerify: false` in the Helm chart causes CI/CD pipelines to detect drift because the API server strips default values, correct?

I'm curious what makes cert-controller different in this regard, since it's merely another thing essentially running kubectl apply. Would cert-controller also be subject to the same default-stripping issue, leaving you right back where you began?

Or is the objective to remove the state from the Helm chart such that there is no longer any check for drift for this value?

Modifying controller code to meet the needs of CI/CD code does seem a bit backwards. I am also wary of potentially manually overriding user intent on what could be a very salient field (is there some reason why insecureSkipTLSVerify is set? Will un-setting it cause an outage? Can we be sure the answer will be identical for all projects that consume cert-controller?)

As for other options, maybe a post-install hook on the Helm chart to modify the value?

Here is an example of a post-install hook G8r uses:

https://github.com/open-policy-agent/gatekeeper/blob/master/charts/gatekeeper/templates/namespace-post-install.yaml

Reply:
> I'm a bit confused here. You're saying that having insecureSkipTLSVerify: false in the Helm chart causes CI/CD pipelines to detect drift because the API server strips default values, correct?

Yes, as it's the default value, it is removed automatically. If you deploy an apiservice with explicitly set `insecureSkipTLSVerify: false` and then you get it from the cluster, you will see that your set value `insecureSkipTLSVerify: false` has been chopped from the manifest. This triggers out-of-sync on CD systems like ArgoCD.
You can just try with this:
```yaml
apiVersion: apiregistration.k8s.io/v1
kind: APIService
metadata:
  labels:
    app.kubernetes.io/name: v1beta1.external.metrics.k8s.io
    app.kubernetes.io/version: latest
    app.kubernetes.io/part-of: keda-operator
  name: v1beta1.custom.metrics.k8s.io
# nosemgrep: yaml.kubernetes.security.skip-tls-verify-service.skip-tls-verify-service
spec:
  service:
    name: keda-metrics-apiserver
    namespace: keda
  group: external.metrics.k8s.io
  version: v1beta1
  groupPriorityMinimum: 100
  versionPriority: 100
  insecureSkipTLSVerify: false
```

> I'm curious what makes cert-controller different in this regard, since it's merely another thing essentially running kubectl apply. Would cert-controller also be subject to the same default-stripping issue, leaving you right back where you began?

Yes, it's removed exactly in the same way, but if it's cert-controller, we don't need to set it in the helm chart, so external tools aren't affected. 

> Or is the objective to remove the state from the Helm chart such that there is no longer any check for drift for this value?

The objective is to not have to explicitly set the value to default value in general. (not only from helm but from raw manifests)

> Modifying controller code to meet the needs of CI/CD code does seem a bit backwards. I am also wary of potentially manually overriding user intent on what could be a very salient field (is there some reason why insecureSkipTLSVerify is set? Will un-setting it cause an outage? Can we be sure the answer will be identical for all projects that consume cert-controller?)

Actually, if cert-controller tries to patch the `apiservice` to add the `caBundle` but the `apiservice` has set `insecureSkipTLSVerify: true`, cert-controller will fail because `insecureSkipTLSVerify: true` and `caBundle` are mutually exclusive. If the value has been set to `true` and then cert-controller has been enabled, cert-controller won't work at all, printing errors all the time due to this mutual exclusivity.

> As for other options, maybe a post-install hook on the Helm chart to modify the value?

We don't need this feature anymore (nor for helm nor for raw yamls) as we just needed it during some releases after the migration from random certs generated by the pods to real certs, stored in Kubernetes and managed by cert-controller/cert-manager. I mean, [we don't need this feature anymore for KEDA](https://github.com/kedacore/charts/pull/564) xD

I've just pushed it instead of just closing the PR because this project is awesome and has helped us a lot and I assume that we aren't the only one who faces with this and the code was already done. 

I can extend any other doubt that you have, said that, if you think that this doesn't make sense, we can close the PR

Reply:
> Actually, if cert-controller tries to patch the apiservice to add the caBundle but the apiservice has set insecureSkipTLSVerify: true, cert-controller will fail because insecureSkipTLSVerify: true and caBundle are mutually exclusive. If the value has been set to true and then cert-controller has been enabled, cert-controller won't work at all, printing errors all the time due to this mutual exclusivity.

This makes sense, but if a user has intentionally skipped TLS verification, will cert-controller overriding it break their use case? If so, consumers would be unable to use that version of cert-controller due to the mandatory override. It comes down to whether its better for cert-controller to return errors or for it to override user intent without context as to why it was set.

> I've just pushed it instead of just closing the PR because this project is awesome and has helped us a lot and I assume that we aren't the only one who faces with this and the code was already done.

I very much appreciate the consideration and am glad the project has helped you!

> I can extend any other doubt that you have, said that, if you think that this doesn't make sense, we can close the PR

I think if the override was optional, such that projects could opt-in, I would be less concerned. This would probably boil down to adding a new option to the rotator:

https://github.com/open-policy-agent/cert-controller/blob/b1a35ebc816cf1ac83a29b1d414da77950967930/pkg/rotator/rotator.go#L219-L259

Which defaults to off so that behavior would be backwards compatible for users on upgrade.

Definitely up to you if the extra work is worth your time. I apologize for the slow review cycle.

Reply:
Making it optional is a good idea indeed! I'll update my PR :) 

>  I apologize for the slow review cycle.

No worries, the time is limited and it's totally normal

Reply:
I've added a parameter to make it optional, but I'm not totally sure about how I can test it (that the parameter isn't removed) without duplicating some code or complicating the tests in general.
I've tried adding a new parameter and propagating it until the validation of `ensureWebhookPopulated` but it requires the propagation across multiple places where it doesn't make sense.
Would it be correct if I just create a single test for testing this specific case directly instead of using current tests?

Reply:
Kindly reminder @maxsmythe 😄

Reply:
I think we probably want to test both cases, but I hear you about not wanting to propagate the expected assertion everywhere.

Maybe we can use functional options to make this a backwards-compatible change to `testWebhook()`.

Functional options blog: https://commandcenter.blogspot.com/2014/01/self-referential-functions-and-design.html

Reply:
f**k! I missed your comment 🤦 
I'll review the PR this week 😄

Reply:
I've added another split test for this specific feature. PTAL when you have some time @maxsmythe

Reply:
Kindly reminder @maxsmythe  😄